src/core/Core.js

```diff
@@ -35,12 +30,181 @@ export default class Core {
 
     this.translator = new Translator({locales: this.opts.locales})
     this.i18n = this.translator.translate.bind(this.translator)
-    // console.log(this.i18n('filesChosen', {smart_count: 3}))
 
     // Set up an event EventEmitter
     this.emitter = new ee.EventEmitter()
 
-    this.totalFilesSelectedCount = 0
+    this.defaultState = {
+      selectedFiles: {},
+      uploadedFiles: {},
+      modal: {
+        isHidden: true,
+        targets: []
+      }
+    }
+
+    this.state = Object.assign({}, this.state, this.defaultState)
+  }
+
+  /**
+   * Iterate on all plugins and run `update` on them. Called when state changes
+   *
+   */
+  updateAll () {
+    Object.keys(this.plugins).forEach((pluginType) => {
+      this.plugins[pluginType].forEach((plugin) => {
+        plugin.update(this.state)
+      })
+    })
+  }
+
+  /**
+   * Reset state to defaultState, used when Modal is closed, for example
+   *
+   */
+  resetState () {
+    this.setState(this.defaultState)
+  }
+
+  /**
+   * Updates state
+   *
+   * @param {newState} object
+   */
+  setState (newState) {
+    this.log(`Update state with: ${newState}`)
+    this.state = Object.assign({}, this.state, newState)
+    this.updateAll()
+  }
+
+  /**
+   * Gets current state, making sure to make a copy of the state object and pass that,
+   * instead of an actual reference to `this.state`
+   *
+   */
+  getState () {
+    return Object.assign({}, this.state)
```

Comment:
> My first thought was someone may think it's the actual state, so they'd call Object.assign on it again

Right, will that cause any issues though? They’d just have an object that has been duplicated twice.

>There may also be an issue if someone fetched the state with getState, then waited for an async callback

That is a problem indeed. We discussed that in Slack with you briefly too. We should study how [other state managing libs](https://github.com/sethvincent/store-emitter/blob/master/index.js) do that, maybe queue state modification calls and do them one by one, or what is the way here. Surely that has been dealt with already.

Somewhat related Kevin’s comment from Slack:
> you’re still mutating the new object, which can be avoided

> I think we can use `deepFreeze` (in our test  / initial phase) which throws an error as soon as you mutate the object that you specify. I saw it in Dan's egghead.io talk on redux last week.